### PR TITLE
Allow playing Animations asynchronously using OpenGLRenderer

### LINKED ISSUE #1267: Starting Animations using Scene.play while others are running
First of all: This might just as well be a bug report - I am unsure...

## Motivation
Look at the following scene:

https://user-images.githubusercontent.com/26899845/113640645-3ab53f80-967c-11eb-8789-4f3832acf0a1.mp4

Swiping a line across the Scene and playing a Flash as soon as it hits a dot... Or more generally: Animate one thing and play additional animations based on the state of the animated thing. Sounds like a job for an update function. Consider the following code:
```py
class UsingUpdater(Scene):
    def construct(self):
        #Setup
        dots = VGroup(*[Dot(RIGHT*random.randrange(-6, 6)+UP*random.randrange(-2, 2)) for i in range(10)])
        line = Line(UP*3, DOWN*3).align_on_border(LEFT)
        self.add(dots)
        self.play(Create(line))

        #Updater
        def flash_hit_dots(mob):
            for dot in dots:
                if dot.get_center()[0] < mob.get_center()[0] and not hasattr(dot, "marked"):
                    self.play(Flash(dot, run_time=0.5))
                    dot.marked = True
        line.add_updater(flash_hit_dots)

        #Animate Line
        self.play(line.animate.align_on_border(RIGHT), run_time=5)
        self.wait()
```
However, this results in `ValueError: write to closed file`.

## Description of proposed feature
The feature request is quite simple, but I think it is hard to implement: Allow using `Scene.play()` while an Animation is running.
 

## Additional comments
I have seen quite a few different workarounds for this problem, even in 3b1b's code. Some are based on splitting the animation at each new event, others are based on precalculating all timings in advance - but all of them have one thing in common: They are tedious and error-prone. In case you are wondering, how I created the example Scene: Have a look at this mess...
```py
class Workaround(Scene):
    def construct(self):
        #Setup
        dots = VGroup(*[Dot(RIGHT*random.randrange(-6, 6)+UP*random.randrange(-2, 2)) for i in range(10)])
        line = Line(UP*3, DOWN*3).align_on_border(LEFT)
        self.add(dots)
        self.play(Create(line))

        run_time = 5
        rate_func = smooth
        fps = config["frame_rate"]
        
        #First pass to determine frames to start the flash:
        animations = []
        frames = run_time*fps
        x_0 = line.get_center()[0]
        x_1 = line.copy().align_on_border(RIGHT).get_center()[0]
        for t in range(frames):
            alpha = rate_func(t/frames)
            x = x_0 + alpha*(x_1-x_0)
            for dot in dots:
                if dot.get_center()[0] < x and not hasattr(dot, "marked"):
                    #create dummy wait animation:
                    wait = FadeOut(Mobject(), run_time=t/fps)
                    animations.append(AnimationGroup(wait, Flash(dot, run_time=0.5), lag_ratio=1))
                    dot.marked = True
        
        #Run the animations
        self.play(AnimationGroup(
            ApplyMethod(line.align_on_border, RIGHT, run_time=run_time, rate_func=rate_func),
            *animations
        ))
        self.wait()
```

### PR DESCRIPTION
**Pull request by AntonBallmaier**
## Changelog / Overview

`Scene.play` can take a new keyword argument `run_async`. If `True` this will start the animation and return to executing `Scene.construct` immediately. This feature is only availably using the `OpenGLRenderer`.

## Motivation
see #1488
closes #1267 

## Explanation for Changes

[Documentation Reference](https://manimce--1526.org.readthedocs.build/en/1526/reference/manim.scene.scene.Scene.html#manim.scene.scene.Scene.play)

## Further Comments
I would like to add an example to the docs, but as far as I know we cannot render OpenGl-videos to the docs, right? 

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [ ] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

**Comment by TRoboto** [comment]
I've tried to play your example #1267, with preview enabled, after converting it to OpenGL, but it didn't work as it should. When I write to a file, I receive this error.
```py
ValueError: write to closed file
```

Find the scene below. Note that you should temporarily change [this line](https://github.com/ManimCommunity/manim/blob/e4bc0ea7d35404910904d4e73b365e278a57e839/manim/animation/indication.py#L258) to `lines = OpenGLVGroup()` in order for it to work.

```py
class UsingUpdater(Scene):
    def construct(self):
        # Setup
        dots = OpenGLVGroup(
            *[
                Dot(RIGHT * random.randrange(-6, 6) + UP * random.randrange(-2, 2))
                for i in range(10)
            ]
        )
        line = Line(UP * 3, DOWN * 3).align_on_border(LEFT)
        self.add(dots)
        self.play(Create(line))

        # Updater
        def flash_hit_dots(mob):
            for dot in dots:
                if dot.get_center()[0] < mob.get_center()[0] and not hasattr(
                    dot, "marked"
                ):
                    self.play(Flash(dot, run_time=0.5))
                    dot.marked = True

        line.add_updater(flash_hit_dots)

        # Animate Line
        self.play(line.animate.align_on_border(RIGHT), run_time=5)
        self.wait()
```

**Comment by AntonBallmaier** [comment]
> I've tried to play your example #1267, with preview enabled, after converting it to OpenGL, but it didn't work as it should. When I write to a file, I receive this error.
> 
> ```python
> ValueError: write to closed file
> ```

That is the error that without the changes I made. Please make sure you are actually running the code on the branch of this PR. I just tested the same code and don't get that error.  

> 
> Find the scene below. Note that you should temporarily change [this line](https://github.com/ManimCommunity/manim/blob/e4bc0ea7d35404910904d4e73b365e278a57e839/manim/animation/indication.py#L258) to `lines = OpenGLVGroup()` in order for it to work.
> 
> ```python
> class UsingUpdater(Scene):
>     def construct(self):
>         # Setup
>         dots = OpenGLVGroup(
>             *[
>                 Dot(RIGHT * random.randrange(-6, 6) + UP * random.randrange(-2, 2))
>                 for i in range(10)
>             ]
>         )
>         line = Line(UP * 3, DOWN * 3).align_on_border(LEFT)
>         self.add(dots)
>         self.play(Create(line))
> 
>         # Updater
>         def flash_hit_dots(mob):
>             for dot in dots:
>                 if dot.get_center()[0] < mob.get_center()[0] and not hasattr(
>                     dot, "marked"
>                 ):
>                     self.play(Flash(dot, run_time=0.5))
>                     dot.marked = True
> 
>         line.add_updater(flash_hit_dots)
> 
>         # Animate Line
>         self.play(line.animate.align_on_border(RIGHT), run_time=5)
>         self.wait()
> ```

I tested the code and had to make a tiny adjustment for it to work:

```py
class UsingUpdater(Scene):
    def construct(self):
        # Setup
        dots = OpenGLVGroup(
            *[
                Dot(RIGHT * random.randrange(-6, 6) + UP * random.randrange(-2, 2))
                for i in range(10)
            ]
        )
        line = Line(UP * 3, DOWN * 3).align_on_border(LEFT)
        self.add(dots)

        # Updater
        def flash_hit_dots(mob):
            for dot in dots:
                if dot.get_center()[0] < line.get_center()[0] and not hasattr(
                    dot, "marked"
                ):
                    self.play(Flash(dot, run_time=0.5, rate_func=rush_from))
                    dot.marked = True

        line.add_updater(flash_hit_dots)

        # Animate Line
        self.play(line.animate.align_on_border(RIGHT), run_time=5)
        self.wait()

```

Resulting in:

https://user-images.githubusercontent.com/26899845/118504901-36288000-b72c-11eb-8c08-c560f057134b.mp4

**Comment by TRoboto** [comment]
The old scene still doesn't work but with your edits, it does. When the `mob` argument is used inside the updater function, everything breaks. Is this supposed to happen?

**Comment by AntonBallmaier** [comment]
> The old scene still doesn't work but with your edits, it does. When the `mob` argument is used inside the updater function, everything breaks. Is this supposed to happen?

Definitely not, and I will open an issue addressing that bug. It is however (as far as I can tell) not related to this PR.
Have a look at this:
```py
class AnimateBugTest(Scene):
    def construct(self):
        line = Line(UP * 3, DOWN * 3).align_on_border(LEFT)
        dot1 = Dot(color=RED)
        dot2 = Dot(color=GREEN)
        self.add(line, dot1, dot2)

        line.add_updater(lambda mob: dot1.move_to(mob))
        dot2.add_updater(lambda mob: mob.move_to(line))

        # Animate Line
        self.play(line.animate.align_on_border(RIGHT), run_time=5)
        self.wait()
```
No matter the renderer the result looks like this:

https://user-images.githubusercontent.com/26899845/118536063-8bc05500-b74b-11eb-9c20-a37478700f94.mp4

File: manim/scene/scene.py [hunk 1]
```diff
@@ -849,8 +851,92 @@ def get_run_time(self, animations):
...
-        self.renderer.play(self, *args, **kwargs)
+    def play(
+        self, *animations: Animation, run_async: Optional[bool] = None, **kwargs: Any
+    ):
+        """Play :class:`Animations<.Animation>` in the scene.
+
+        Parameters
+        ----------
+        animations
+            The animations to be played. They are started at the same timed and played
```
Comment (TRoboto) [inline manim/scene/scene.py:862]: timed is a typo here.

File: manim/scene/scene.py [hunk 2]
```diff
@@ -946,6 +1032,7 @@ def play_internal(self, skip_rendering=False):
             named parameters affecting what was passed in ``args``,
             e.g. ``run_time``, ``lag_ratio`` and so on.
         """
+        self.sync_play_running = True
```
Comment (TRoboto) [inline manim/scene/scene.py:1035]: Is there a reason to set assign this variable here? Will I have to call wait for the async to work?

**Comment by TRoboto** [comment]
> > The old scene still doesn't work but with your edits, it does. When the `mob` argument is used inside the updater function, everything breaks. Is this supposed to happen?
> 
> Definitely not, and I ~will open~ have opened an issue addressing that bug. It is however ~(as far as I can tell)~ (for sure) not related to this PR. See #1536 :) Thanks for the tip making me look into this +1

Thank you for the unstoppable contributions, your work is astonishing. I've added few comments for your consideration.

**Comment by TRoboto** [comment]
> Definitely not, and I ~will open~ have opened an issue addressing that bug. It is however ~(as far as I can tell)~ (for sure) not related to this PR. See #1536 :) Thanks for the tip making me look into this +1

Thank you for the unstoppable contributions, your work is astonishing. This PR LGTM now.

**Review by huguesdevimeux — CHANGES_REQUESTED** [review]
So, that's a great addition. I left some comments. 
This crucially needs two things to my eyes : 
- Comments of the logic  ; there are parts that are kind of obscure and need explanation. 
- This needs test, I think both graphical and units. You could use mocks to test, I think. 

Aside, that's (again!) a great contribution !

**Comment by Darylgolden** [comment]
Hi, can I check on the status of this PR?

**Comment by MrDiver** [comment]
Closing this for now because there is no activity. We can come back to this. The issue is on the project board and the PR is linked!

**Pull request closed without merging**